core/src/main/java/kafka/server/ReplicaFetcherTierStateMachine.java

```diff
@@ -238,12 +238,8 @@ private Long buildRemoteLogAuxState(TopicPartition topicPartition,
 
                 log.debug("Updated the epoch cache from remote tier till offset: {} with size: {} for {}", leaderLocalLogStartOffset, epochs.size(), partition);
 
-                // Restore producer snapshot
-                File snapshotFile = LogFileUtils.producerSnapshotFile(unifiedLog.dir(), nextOffset);
-                buildProducerSnapshotFile(snapshotFile, remoteLogSegmentMetadata, rlm);
-
                 // Reload producer snapshots.
-                unifiedLog.producerStateManager().truncateFullyAndReloadSnapshots();
+                truncateFullyAndReloadRestoredSnapshots(unifiedLog, nextOffset, remoteLogSegmentMetadata, rlm);
```

Comment:
@hudeqi 

> In producerStateManager, first call truncateFullyAndStartAt to clean up the snapshot files, and then pull the snapshot file from RemoteLogManager. If the order of calls is reversed in original logic, it may cause the newly built snapshot file to be cleaned up again by truncateFullyAndStartAt.

The `truncateFullyAndReloadSnapshots` only removes the snapshot files that was already loaded into the ProducerStateManager, so it doesn't remove the `snapshotFile` that was downloaded/built from the remote storage.

Reply:
> @hudeqi
> 
> > In producerStateManager, first call truncateFullyAndStartAt to clean up the snapshot files, and then pull the snapshot file from RemoteLogManager. If the order of calls is reversed in original logic, it may cause the newly built snapshot file to be cleaned up again by truncateFullyAndStartAt.
> 
> The `truncateFullyAndReloadSnapshots` only removes the snapshot files that was already loaded into the ProducerStateManager, so it doesn't remove the `snapshotFile` that was downloaded/built from the remote storage.
> 
> AFAIR, we don't want to expose reloading the snapshots from ProducerStateManager without clearing it's internal states (`ProducerStateManager#reloadSnapshots` method). We can add a comment int this method to capture this information for future readers.

Hi, @kamalcph , the reason why such changes are made in ReplicaFetcherTierStateMachine here is because of a relatively coincidental boundary case: if the name of the file pulled from the remote storage is the same as the name of the local file before truncate (that is, the file name pulled from the remote storage is in the keySet of the `snapshots` map), the snapshot file built first will be cleared by the later `truncateFullyAndStartAt`. 

I was adding the current unit test and found that the test was failed, so I discovered this issue.

Reply:
Not sure when this case can happen in the leader-follower fetch protocol. We can keep the refactoring if required to maintain the code clarity. 

Can we update the title since this is not an issue?

Reply:
updated.